File: leetcode/medium/16-3sum-closest.cpp
#include <bits/stdc++.h>
using namespace std;


class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());

        int numsLen = nums.size();
        int res = INT_MAX - 1000;
        for (int i=0; i<numsLen-2; i++) {
            if (i>0 && nums[i] == nums[i-1]) continue;

            int low = i + 1;
            int high = numsLen - 1;
            while (high > low) {
                int sum = nums[i] + nums[low] + nums[high];
                
                if (sum == target) {
                    return sum;
                }
                else if (sum < target) low++;
                else high--;

                int prevDist = abs(target - res);
                int curtDist = abs(target - sum);
                if (prevDist > curtDist) {
                    res = sum;
                }
            }
        }
        return res;
    }
};


int main() {
    Solution solve;

    vector<int> nums = {-1,2,1,-4};
    int target = 1;

    int res = solve.threeSumClosest(nums, target);
    cout << res << endl;

    return 0;
}

File: leetcode/medium/16-3sum-closest.py
class Solution:
    def threeSumClosest(self, nums, target):
        nums.sort()
        
        numsLen = len(nums)
        res = float('inf')
        for i in range(0, numsLen-2):
            if i > 0 and (nums[i] == nums[i-1]): continue
            
            low = i + 1
            high = numsLen - 1
            while high > low:
                sum = nums[i] + nums[low] + nums[high]
                
                if sum == target: return sum
                elif sum < target: low += 1
                else: high -= 1
                
                prevDistance = abs(target - res)
                currDistance = abs(target - sum)
                if (prevDistance > currDistance):
                    res = sum
        return res


solve = Solution()

nums = [-1,2,1,-4]
target = 1

res = solve.threeSumClosest(nums, target)
print(res)
